File: 210-CourseScheduleIi/210-CourseScheduleIi.cpp
// Last updated: 17/07/2025, 20:52:24
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> adj[numCourses];
        for(auto it: prerequisites){
            adj[it[1]].push_back(it[0]);
        }
        int inorder[numCourses] ;
        for(int i = 0;i<numCourses;i++){
            inorder[i] = 0;
        }
        for(int i = 0; i<numCourses;i++){
            for(auto it: adj[i]){
                inorder[it]++;
            }
        }
        // for(int i = 0;i<numCourses;i++){
        //     cout<<inorder[i]<<" ";
        // }
        queue<int> q;
        vector<int> ans;
        for(int i = 0; i< numCourses;i++){
            if(inorder[i]==0){
                q.push(i);
                
            }
        }
        
        while(!q.empty()){
            int node = q.front();
            ans.push_back(node);
            q.pop();
            for(auto it: adj[node]){
                inorder[it]--;
                if(inorder[it]==0){
                    q.push(it);
                    // ans.push_back(it);
                }
            }
            
        }
        if(ans.size()==numCourses)
        return ans;
        return {};

    }
};

File: 210-CourseScheduleIi/210-CourseScheduleIi.py
# Last updated: 01/10/2025, 00:13:16
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = {i : [] for i in range(numCourses)}
        indegree = [0]* numCourses
        for course, pre in prerequisites:
            adj[pre].append(course)
            indegree[course] += 1
        queue = deque([])
        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)
        order = []

        while queue:
            node = queue.popleft()
            order.append(node)
            for neighbor in adj[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)
        
        if len(order) == numCourses:
            return order
        else:
            return []